### PR TITLE
mock-classifier push image to ghcr and sign the image

### PR DESCRIPTION
**Pull request by keijoraamat**
This PR is resolving 

- [https://github.com/buerokratt/Mock-Classifier/issues/31](https://github.com/buerokratt/Mock-Classifier/issues/31)
- [https://github.com/buerokratt/Mock-Classifier/issues/23](https://github.com/buerokratt/Mock-Classifier/issues/23)

### CODE REVIEW

File: .github/workflows/ci-pullrequest-main.yml [hunk 1]
```diff
@@ -79,12 +79,44 @@ jobs:
...
 
     steps:
     - uses: actions/checkout@v3
     
     - name: Docker Setup BuildX
       uses: docker/setup-buildx-action@v2
+
+    - name: Set image tag
+      run: |
+        echo "DOCKER_TAG_SHA=ghcr.io/buerokratt/mock-classifier:${GITHUB_SHA}" >> $GITHUB_ENV
```
Comment (yahya130) [inline .github/workflows/ci-pullrequest-main.yml:95]: Could we set these environment variables under the env section like?
PackageDeploy:
    runs-on: ubuntu-latest
    .....
    env:
      DOCKER_TAG_SHA: ghcr.io/buerokratt/mock-classifier:${GITHUB_SHA}
      DOCKER_TAG_LATEST: ghcr.io/buerokratt/mock-classifier:latest

File: .github/workflows/ci-pullrequest-main.yml [hunk 2]
```diff
@@ -79,12 +79,44 @@ jobs:
...
+        echo "DOCKER_TAG_LATEST=ghcr.io/buerokratt/mock-classifier:latest" >> $GITHUB_ENV
```
Comment (yahya130) [inline .github/workflows/ci-pullrequest-main.yml:96]: This variable is not used anywhere in the file, could it be removed?

File: .github/workflows/ci-pullrequest-main.yml [hunk 3]
```diff
@@ -79,12 +79,44 @@ jobs:
...
     
     - name: Docker Build
-      run: docker image build src/MockClassifier.Api/
+      run: docker image build src/MockClassifier.Api/ --tag $DOCKER_TAG_SHA
```
Comment (yahya130) [inline .github/workflows/ci-pullrequest-main.yml:99]:
```suggestion
      working-directory: src/MockClassifier.Api
      run: docker image build . --tag $DOCKER_TAG_SHA
```
would be a 'nice-to-have' :)

File: .github/workflows/ci-pullrequest-main.yml [hunk 4]
```diff
@@ -79,12 +79,44 @@ jobs:
...
+    - name: List Docker images
+      run: docker images
+
+    - name: Install cosign
+      uses: sigstore/cosign-installer@d6a3abf1bdea83574e28d40543793018b6035605
+      with:
+        cosign-release: 'v1.7.1'
+
+    - name: Log in to GitHub container registry
+      run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login ghcr.io -u $ --password-stdin
```
Comment (martinkearn) [inline .github/workflows/ci-pullrequest-main.yml:110]: @keijoraamat as part of the PR, can you add some guidance to the `readme.md` on how to setup the environment vars or at least cover which vars are needed?

**Review by martinkearn — APPROVED** [review]

File: .github/workflows/ci-pullrequest-main.yml (hunk 1 above)
Comment (keijoraamat) [reply .github/workflows/ci-pullrequest-main.yml:95]: Not all variables can be set at env section at the begging. The problem is that some variables don't have value what are used to create other env values at this stage. For example `$GITHUB_SHA` gets its value after the repo is cloned thus making hard to create tag for docker image with the latest commit SHA.

File: .github/workflows/ci-pullrequest-main.yml (hunk 2 above)
Comment (keijoraamat) [reply .github/workflows/ci-pullrequest-main.yml:96]: Good point. I'll fix that.

File: .github/workflows/ci-pullrequest-main.yml (hunk 4 above)
Comment (keijoraamat) [reply .github/workflows/ci-pullrequest-main.yml:109]: The GITHUB_TOKEN is a special env var. It's set by GitHub and one can control it through `permissions:` part of the job. Did you meant that or something else? It's just that for me, it's quite clear and I don't know what exactly to write down. I would love to explain though.

Comment (martinkearn) [reply .github/workflows/ci-pullrequest-main.yml:107]: Let's discuss it

Comment (martinkearn) [reply .github/workflows/ci-pullrequest-main.yml:107]: Having discussed with @keijoraamat, it seems that `secrets.GITHUB_TOKEN` is built into GitHub so no human intervention required.

**Review by yahya130 — APPROVED** [review]

**Comment by keijoraamat** [comment]
Moved building, pushing to registry and signing to a separate pipeline. Because in the PR latest commit gets squashed together with all commits and merged into `main` branch and therefore this PR-s commit will disappear there won't be a commit from the docker image is builded. So building and publishing takes place after the PR is merged.

**Pull request merged**